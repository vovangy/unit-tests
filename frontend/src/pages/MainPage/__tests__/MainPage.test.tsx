import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MainPage } from '../MainPage';
import { useCurrentTime, useProducts } from '../../../hooks';
import { applyCategories, updateCategories } from '../../../utils';

jest.mock('../../../utils');
jest.mock('../../../hooks');

describe('Main Page tests', () => {
    it('should render correctly', () => {
        jest.mocked(applyCategories).mockReturnValue([
            {
                id: 1,
                name: 'IPhone 14 Pro',
                description: 'Последний iphone, купи его сейчас',
                price: 999,
                priceSymbol: '$',
                category: 'Электроника',
                imgUrl: '/iphone.png',
            },
            {
                id: 2,
                name: 'Костюм гуся',
                description: 'Запускаем гуся, работяги',
                price: 1000,
                priceSymbol: '₽',
                category: 'Одежда',
            },
        ]);
        jest.mocked(useCurrentTime).mockReturnValue('12:00:00');
        const rendered = render(<MainPage />);
        expect(rendered).toMatchSnapshot();
    });

    it('should use useCurrentTime', () => {
        render(<MainPage />);
        expect(useCurrentTime).toHaveBeenCalled();
    });

    it('should use useProducts', () => {
        render(<MainPage />);
        expect(useProducts).toHaveBeenCalled();
    });

    it('should update categories when clicked', () => {
        jest.mocked(updateCategories).mockImplementation(
            (selected, clicked) => [...selected, clicked]
        );

        const rendered = render(<MainPage />);
        const categoryBtn = screen.getByText('Электроника', {
            selector: '.categories__badge',
        });

        fireEvent.click(categoryBtn);

        expect(updateCategories).toHaveBeenCalledWith([], 'Электроника');
        expect(
            screen.getByText('Электроника', {
                selector: '.categories__badge',
            })
        ).toHaveClass('categories__badge_selected');
    });
});

