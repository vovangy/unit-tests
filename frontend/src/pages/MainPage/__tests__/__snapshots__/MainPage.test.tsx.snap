// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Page tests should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="main-page"
      >
        <h2
          class="main-page__title"
        >
          VK Маркет
        </h2>
        <h3>
          12:00:00
        </h3>
        <div
          class="main-page__parameters"
        >
          <div
            class="categories"
          >
            <div
              class="categories__badge"
            >
              Одежда
            </div>
            <div
              class="categories__badge"
            >
              Для дома
            </div>
            <div
              class="categories__badge"
            >
              Электроника
            </div>
          </div>
        </div>
        <div
          class="product-card"
        >
          <div
            class="product-card__text"
          >
            <h2
              class="product-card__name"
            >
              IPhone 14 Pro
            </h2>
            <div
              class="product-card__description"
            >
              Последний iphone, купи его сейчас
            </div>
            <h1
              class="product-card__price"
            />
            <div
              class="product-card__category"
            >
              Электроника
            </div>
          </div>
          <img
            alt="IPhone 14 Pro"
            class="product-card__image"
            src="/iphone.png"
          />
        </div>
        <div
          class="product-card"
        >
          <div
            class="product-card__text"
          >
            <h2
              class="product-card__name"
            >
              Костюм гуся
            </h2>
            <div
              class="product-card__description"
            >
              Запускаем гуся, работяги
            </div>
            <h1
              class="product-card__price"
            />
            <div
              class="product-card__category"
            >
              Одежда
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="main-page"
    >
      <h2
        class="main-page__title"
      >
        VK Маркет
      </h2>
      <h3>
        12:00:00
      </h3>
      <div
        class="main-page__parameters"
      >
        <div
          class="categories"
        >
          <div
            class="categories__badge"
          >
            Одежда
          </div>
          <div
            class="categories__badge"
          >
            Для дома
          </div>
          <div
            class="categories__badge"
          >
            Электроника
          </div>
        </div>
      </div>
      <div
        class="product-card"
      >
        <div
          class="product-card__text"
        >
          <h2
            class="product-card__name"
          >
            IPhone 14 Pro
          </h2>
          <div
            class="product-card__description"
          >
            Последний iphone, купи его сейчас
          </div>
          <h1
            class="product-card__price"
          />
          <div
            class="product-card__category"
          >
            Электроника
          </div>
        </div>
        <img
          alt="IPhone 14 Pro"
          class="product-card__image"
          src="/iphone.png"
        />
      </div>
      <div
        class="product-card"
      >
        <div
          class="product-card__text"
        >
          <h2
            class="product-card__name"
          >
            Костюм гуся
          </h2>
          <div
            class="product-card__description"
          >
            Запускаем гуся, работяги
          </div>
          <h1
            class="product-card__price"
          />
          <div
            class="product-card__category"
          >
            Одежда
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
